src/reducers/versions.tsx

```diff
@@ -1157,11 +1162,15 @@ const reducer: Reducer<VersionsState, ActionType<typeof actions>> = (
     case getType(actions.loadVersionInfo): {
       const { version } = action.payload;
 
+      const expandedPaths =
+        state.expandedPaths || getParentFolders(version.file.selected_file);
```

Comment:
I added the filtering out here, and some tests for this. I realized it may be a bit more tricky than I thought... When switching to a version which has been loaded before, `loadVersionInfo` will not be dispatched, so the paths that do not exist in new version cannot be removed unless we introduce a new action to do this. On the other hand, it looks not harmful if we just leave those nonexistent paths there (or I missed something?).

Reply:
Would it help to fix the paths when `setCurrentVersionId` gets dispatched?

https://github.com/mozilla/addons-code-manager/blob/3afafcd01ce6051f7cff6bc1dbee9f02ab021828/src/reducers/versions.tsx#L1459-L1465

I'm not sure how well `setCurrentVersionId` is supported in places like the compare page, tbh.

You could be right, maybe it doesn't matter if there are foreign paths in `expandedPaths`. My one concern is that switching between lots of versions will start to introduce a memory leak. However, a fix for that could be [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it).

Reply:
> Would it help to fix the paths when setCurrentVersionId gets dispatched?

Great. It looks like a good place to fix `expandedPath`. I come up with the following patch, which moves path filtering into `setCurrentVersionId`. It is a bit strange to let `setCurrentVersionId` do such clean-up, but it works. Do you want me to proceed in this direction?
```diff
diff --git a/src/reducers/versions.tsx b/src/reducers/versions.tsx
index a437f60..abdbad0 100644
--- a/src/reducers/versions.tsx
+++ b/src/reducers/versions.tsx
@@ -1164,9 +1164,8 @@ const reducer: Reducer<VersionsState, ActionType<typeof actions>> = (
       const { expandedPaths: paths } = state;
       const entries = Object.keys(version.file.entries);
 
-      const expandedPaths = paths
-        ? paths.filter((e) => entries.includes(e))
-        : getParentFolders(version.file.selected_file);
+      const expandedPaths =
+        paths || getParentFolders(version.file.selected_file);
 
       return {
         ...state,
@@ -1483,6 +1482,26 @@ const reducer: Reducer<VersionsState, ActionType<typeof actions>> = (
     }
     case getType(actions.setCurrentVersionId): {
       const { versionId } = action.payload;
+      let { expandedPaths } = state;
+      const version = getVersionInfo(state, versionId);
+
+      if (version && expandedPaths) {
+        const paths = version.entries.map((e) => e.path);
+        expandedPaths = expandedPaths.filter((e) => paths.includes(e));
+
+        return {
+          ...state,
+          currentVersionId: versionId,
+          expandedPaths,
+          versionInfo: {
+            [versionId]: {
+              ...version,
+              expandedPaths,
+            },
+          },
+        };
+      }
+
       return {
         ...state,
         currentVersionId: versionId,

```

Reply:
Yeah, I think it's worth proceeding. I double checked and it looks like `setCurrentVersionId()` is in fact dispatched in all the right places. I wouldn't know for sure until trying things out, though.

Reply:
I wouldn't worry about supporting `versionInfo` in this new reducer because we don't need to invest in new features for the `version.expandedPaths` property.